include/universal/blas/generators/randsvd.hpp

```diff
@@ -11,54 +13,54 @@ namespace sw {
     namespace universal {
         namespace blas {
 
-            template<typename T>
-            int find_rank(vector<vector<T>> A) {
-                int n = num_rows(A);
-                int m = num_cols(A);
+            template<typename Scalar>
+            size_t find_rank(const matrix<Scalar> A) {
+                size_t n = num_rows(A);
+                size_t m = num_cols(A);
 
-                int rank = 0;
+                size_t rank = 0;
                 vector<bool> row(n, false);
-                for (T i = 0; i < m; ++i) {
-                    for (T j = 0; j < n; ++j) {
+                for (size_t i = 0; i < m; ++i) {
+                    for (size_t j = 0; j < n; ++j) {
                         if (abs(A[j][i]) > EPS && !row[j]) break;
                     }
 
                     if (j != n) {
                         ++rank;
                         row[j] = true;
-                        for (T p = i + 1; p < m; ++p) A[j][p] /= A[j][i];
-                        for (T k = 0; k < n; ++k) {
+                        for (size_t p = i + 1; p < m; ++p) A[j][p] /= A[j][i];
+                        for (size_t k = 0; k < n; ++k) {
                             if (abs(A[k][i]) > EPS && k != j) {
-                                for (T p = i + 1; p < m; ++p)
+                                for (size_t p = i + 1; p < m; ++p)
                                     A[k][p] -= A[j][p] * A[k][i];
                             }
                         }
                     }
                 }
                 return rank;
             }
-
-            template<typename T>
-            boost::tuple randsvd(vector<vector<T>> A) {
-                int k = min(num_cols(A), num_rows(A));
-                int n = num_cols(A), m = num_row(A);
-                //generate a gaussian random matrix of size n x k 
-                vector<vector<T>> omega(n, vector<T>(k));
-                vector<vector<T>> Y(m, vector<T>(k));
-                vector<vector<T>> Q(m, vector<T>(k));
-                vector<vector<T>> R(m, vector<T>(k));
-                vector<vector<T>> B(n, vector<T>(k));
-                vector<vector<T>> S(n, vector<T>(n));
-                vector<vector<T>> V(n, vector<T>(k));
-                vector<vector<T>> D(k, vector<T>(k));
+            template<typename Scalar>
+            void qr(const matrix<Scalar> A, matrix<Scalar> Q, matrix<Scalar> R) {
+            	//R is the upper triangular matrix
+            	//Q is the orthogonal matrix
+            	
+            }
+            
+            template<typename Scalar>
+            tuple randsvd(const matrix<Scalar> &A) {

```

Comment:
personally, I like to put the reference symbol next to the type, not the variable, as in:
tuple randsvd(const matrix<Scalar>&      A) { ... }
As the & symbol modifies the type, it should be associated visually with the type it is modifying.